src/s3ql/backends/b2/b2_backend.py

```diff
@@ -0,0 +1,807 @@
+        self.b2_application_key = options.backend_password
+
+        self.tcp_timeout = self.options.get('tcp-timeout', 20)
+
+        self.account_id = self.options.get('account-id', None)
+
+        self.disable_versions = self.options.get('disable-versions', False)
+        self.retry_on_cap_exceeded = self.options.get('retry-on-cap-exceeded', False)
+
+        # Test modes
+        self.test_mode_fail_some_uploads = self.options.get('test-mode-fail-some-uploads', False)
+        self.test_mode_expire_some_tokens = self.options.get('test-mode-expire-some-tokens', False)
+        self.test_mode_force_cap_exceeded = self.options.get('test-mode-force-cap-exceeded', False)
+
+        (bucket_name, prefix) = self._parse_storage_url(options.storage_url, self.ssl_context)
+        self.bucket_name = bucket_name
+        self.bucket_id = None
+        self.prefix = prefix
+
+        self.api_version = 2
+        self.api_url_prefix = '/b2api/v' + str(self.api_version) + '/'
+
+        # are set by _authorize_account
+        self.api_url = None
+        self.download_url = None
+        self.authorization_token = None
+        self.api_connection = None
+        self.download_connection = None
+
+        self.info_header_prefix = 'X-Bz-Info-'
+
+    def _get_key_with_prefix(self, key):
+        return self._b2_url_encode('%s%s' % (self.prefix, key))
+
+    @staticmethod
+    def _parse_storage_url(storage_url, ssl_context):
+        '''Extract information from storage URL
+
+        Return a tuple * (bucket_name, prefix) * .
+        '''
+
+        hit = re.match(r'^b2://([^/]+)(?:/(.*))?$', storage_url)
+
+        if not hit:
+            raise QuietError('Invalid storage URL', exitcode=2)
+
+        bucket_name = hit.group(1)
+
+        # Backblaze has bucket name restrictions:
+        if not re.match(r'^(?!b2-)[a-z0-9A-Z\-]{6,50}$', bucket_name):
+            raise QuietError('Invalid bucket name.', exitcode=2)
+
+        prefix = hit.group(2) or ''
+        if prefix != '':
+            # add trailing slash if there is none
+            prefix = prefix + '/' if prefix[-1] != '/' else prefix
+
+        return (bucket_name, prefix)
+
+    def _reset_authorization_values(self):
+        self.api_url = None
+        self.download_url = None
+        self.authorization_token = None
+        self._close_connections()
+
+    def _close_connections(self):
+        if self.api_connection:
+            self.api_connection.disconnect()
+            self.api_connection = None
+
+        if self.download_connection:
+            self.download_connection.disconnect()
+            self.download_connection = None
+
+    def _reset_connections(self):
+        if self.api_connection:
+            self.api_connection.reset()
+
+        if self.download_connection:
+            self.download_connection.reset()
+
+    def _get_download_connection(self):
+        if self.download_connection is None:
+            self._authorize_account()
```

Comment:
The authorization request returns `downloadUrl` and `apiUrl` which are then used to create the connection. So, if there is no connection it gets created by calling `self._authorize_account()`. But I suppose I can extract this into separate functions to avoid calling the API again. But this case should actually never happen because both connection should be created and used until the authorization token expires.

Reply:
You are calling `_close_connections` in `open_read`'s error case. That would benefit from not re-authorising.